pkg/metadata/store/store.go

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package store
+
+// All byte arrays are expected to be read-only. User MUST NOT modify byte
+// array element directly!!
+
+// UpdateFunc is function used to update a specific metadata. The value
+// passed in is the current value. The updated value should be returned.
+// If there is an error, the update will be rolled back.
+type UpdateFunc func([]byte) ([]byte, error)
+
+// Metadata is a metadata in the store. It supports transactional update.
+type Metadata interface {
+	// Get a snapshot of the metadata.
+	Get() ([]byte, error)
+	// Update the value.
+	Update(UpdateFunc) error
+}
+
+// MetadataStore is the interface for storing metadata.
+// TODO(random-liu): Initialize the metadata store with a type, and replace
+// []byte with interface{}, so as to avoid extra marshal/unmarshal on the
+// user side.
+type MetadataStore interface {
+	// Create the metadata with the specified id.
+	// Note:
+	// * Create MUST return error if the id already exists.
+	// * The id and metadata MUST be added in one transaction.
+	Create(string, []byte) (Metadata, error)
+	// Get the metadata by id.
+	Get(string) (Metadata, error)
+	// List all the metadatas.
+	List() ([]Metadata, error)
+	// Delete the metadata by id.
+	// Note:
+	// * Delete MUST not return error if the id doesn't exist.
```

Comment:
I see so:
// Delete(id) returns error = nil when there is no metadata for the id or when the metadata has been successfully deleted. Otherwise Delete(id) returns an error !=nil when there is metadata but the delete operation has failed...